go-controller/hybrid-overlay/pkg/windows/service/service.go

```diff
@@ -35,42 +34,46 @@ import (
 )
 
 type handler struct {
-	fromsvc chan error
+	svcName string
+	ctx     *context.Context
+	err     chan error
 }
 
-// InitService is the entry point for running the daemon as a Windows
-// service. It returns an indication of whether it is running as a service;
-// and an error.
-func InitService(serviceName string) error {
+func InitService(serviceName string, ctx *context.Context) error {
 	h := &handler{
-		fromsvc: make(chan error),
+		svcName: serviceName,
+		ctx:     ctx,
+		err:     make(chan error),
 	}
 
 	var err error
 	go func() {
 		err = svc.Run(serviceName, h)
-		h.fromsvc <- err
+		h.err <- err
```

Comment:
@mansikulkarni96 this is a bit fishy, because it's not obvious who is going to be writing into the channel (either this goroutine returns an error from svc.Run() or the Execute() handler returns nil, and the channel is used as a synchronization point. Here's what I'd do instead using a WaitGroup to make synchronization clearer:

```
func InitService(serviceName string, ctx *context.Context) error {
	h := &handler{
		svcName: serviceName,
		ctx:     ctx,
		wg:      &sync.WaitGroup{},
	}

	var err error
	h.wg.Add(1)
	go func() {
		err = svc.Run(serviceName, h)
		if err != nil {
			// Failed before handler started executing
			h.wg.Done()
		}
	}()
	h.wg.Wait()

	return err
}

func (h *handler) Execute(_ []string, r <-chan svc.ChangeRequest, s chan<- svc.Status) (bool, uint32) {
	s <- svc.Status{State: svc.StartPending, Accepts: 0}
	// Unblock initService()
	h.wg.Done()

	ctx, cancel := context.WithCancel(*h.ctx)
```